File: SS/Labs/Lab7/lr0_lab7.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define STATES 7
#define TERMS 3      /* a, b, $ */
#define NONTERMS 2   /* E, T */
#define MAX_STACK 256

/* Productions:
   (1) E -> T T
   (2) T -> a T
   (3) T -> b
*/
static const char lhs[] = {'?', 'E', 'T', 'T'};
static const int rhs_len[] = {0, 2, 2, 1};
static const char *prod_text[] = {
    "",
    "E -> T T",
    "T -> a T",
    "T -> b"
};

/* LR(0) ACTION table encoded as strings: sX, rY, acc, - */
static const char *action[STATES][TERMS] = {
    {"s3", "s4", "-"},
    {"-", "-", "acc"},
    {"s3", "s4", "-"},
    {"s3", "s4", "-"},
    {"r3", "r3", "r3"},
    {"r1", "r1", "r1"},
    {"r2", "r2", "r2"}
};

/* GOTO[E, T] */
static const int gotos[STATES][NONTERMS] = {
    {1, 2},
    {-1, -1},
    {-1, 5},
    {-1, 6},
    {-1, -1},
    {-1, -1},
    {-1, -1}
};

static int term_index(char c) {
    if (c == 'a') return 0;
    if (c == 'b') return 1;
    if (c == '$' || c == '\0') return 2;
    return -1;
}

static int nonterm_index(char c) {
    if (c == 'E') return 0;
    if (c == 'T') return 1;
    return -1;
}

static void print_table(void) {
    int i;
    printf("\nLR(0) Parse Table\n");
    printf("State |   a   b   $  ||  E   T\n");
    printf("--------------------------------\n");
    for (i = 0; i < STATES; i++) {
        printf("  %d   | %-3s %-3s %-3s || ", i, action[i][0], action[i][1], action[i][2]);
        if (gotos[i][0] >= 0) printf("%-3d ", gotos[i][0]); else printf("-   ");
        if (gotos[i][1] >= 0) printf("%-3d", gotos[i][1]); else printf("-  ");
        printf("\n");
    }
}

static void print_stack(const int st[], const char sym[], int top, char *out, size_t n) {
    size_t used = 0;
    int i;
    used += snprintf(out + used, n - used, "%d", st[0]);
    for (i = 1; i <= top && used < n; i++) {
        used += snprintf(out + used, n - used, " %c %d", sym[i], st[i]);
    }
}

static int parse(const char *raw) {
    char input[256];
    int st[MAX_STACK];
    char sym[MAX_STACK];
    int top = 0;
    int ip = 0;
    int step = 1;

    strncpy(input, raw, sizeof(input) - 2);
    input[sizeof(input) - 2] = '\0';
    strcat(input, "$");

    st[0] = 0;
    sym[0] = '#';

    printf("\nParser Moves (LR(0)) for input \"%s\"\n", raw);
    printf("Step | Stack                     | Input     | Action\n");
    printf("-----------------------------------------------------------\n");

    while (1) {
        int state = st[top];
        int ti = term_index(input[ip]);
        char stack_view[256];
        const char *act;

        if (ti < 0) {
            print_stack(st, sym, top, stack_view, sizeof(stack_view));
            printf("%-4d | %-25s | %-9s | error (invalid symbol)\n", step, stack_view, input + ip);
            return 0;
        }

        act = action[state][ti];
        print_stack(st, sym, top, stack_view, sizeof(stack_view));

        if (act[0] == 's') {
            int next = atoi(act + 1);
            printf("%-4d | %-25s | %-9s | shift %d\n", step, stack_view, input + ip, next);
            if (top + 1 >= MAX_STACK) return 0;
            top++;
            sym[top] = input[ip++];
            st[top] = next;
        } else if (act[0] == 'r') {
            int p = atoi(act + 1);
            int pop_count = rhs_len[p];
            int base_state;
            int g;
            printf("%-4d | %-25s | %-9s | reduce %s\n", step, stack_view, input + ip, prod_text[p]);
            if (top - pop_count < 0) return 0;
            top -= pop_count;
            base_state = st[top];
            g = gotos[base_state][nonterm_index(lhs[p])];
            if (g < 0 || top + 1 >= MAX_STACK) return 0;
            top++;
            sym[top] = lhs[p];
            st[top] = g;
        } else if (strcmp(act, "acc") == 0) {
            printf("%-4d | %-25s | %-9s | accept\n", step, stack_view, input + ip);
            return 1;
        } else {
            printf("%-4d | %-25s | %-9s | error\n", step, stack_view, input + ip);
            return 0;
        }

        step++;
    }
}

int main(void) {
    char input[256];

    printf("Grammar:\n");
    printf("E -> T T\n");
    printf("T -> a T | b\n");

    print_table();

    printf("\nEnter input string over {a,b}: ");
    if (!fgets(input, sizeof(input), stdin)) return 0;

    if (strlen(input) > 0 && input[strlen(input) - 1] == '\n') {
        input[strlen(input) - 1] = '\0';
    }

    printf("\nResult: %s\n", parse(input) ? "Accepted" : "Rejected");
    return 0;
}

File: SS/Labs/PythonLabs/Lab7/lr0_lab7.py
# Lab 7 — LR(0) Parser
# Grammar:
#   (1) E -> T T
#   (2) T -> a T
#   (3) T -> b

prods   = [None, ('E', 'TT'), ('T', 'aT'), ('T', 'b')]
rhs_len = [0,    2,            2,            1]

# LR(0) ACTION table: 'sX' = shift to X, 'rY' = reduce by Y, 'acc', '-' = error
action = [
    #  a      b      $
    ['s3',  's4',  '-'  ],   # 0
    ['-',   '-',   'acc'],   # 1
    ['s3',  's4',  '-'  ],   # 2
    ['s3',  's4',  '-'  ],   # 3
    ['r3',  'r3',  'r3' ],   # 4
    ['r1',  'r1',  'r1' ],   # 5
    ['r2',  'r2',  'r2' ],   # 6
]

# GOTO[state][E=0, T=1]
gotos = [
    [1, 2], [-1, -1], [-1, 5], [-1, 6],
    [-1, -1], [-1, -1], [-1, -1]
]

TERMS    = ['a', 'b', '$']
NONTERMS = ['E', 'T']

def term_idx(c):
    if c == 'a':  return 0
    if c == 'b':  return 1
    if c in ('$', ''): return 2
    return -1

def nonterm_idx(c):
    if c == 'E': return 0
    if c == 'T': return 1
    return -1

def print_table():
    print("\nLR(0) Parse Table")
    print("State |   a   b   $  ||  E   T")
    print("--------------------------------")
    for i, row in enumerate(action):
        g = gotos[i]
        ge = str(g[0]) if g[0] >= 0 else '-'
        gt = str(g[1]) if g[1] >= 0 else '-'
        print(f"  {i}   | {row[0]:<3} {row[1]:<3} {row[2]:<3} || {ge:<3} {gt}")

def parse(raw):
    inp = raw + '$'
    st   = [0]
    sym  = ['#']
    ip   = 0
    step = 1

    print(f'\nParser Moves (LR(0)) for input "{raw}"')
    print(f"{'Step':<5} {'Stack':<27} {'Input':<10} Action")
    print("-" * 60)

    while True:
        state = st[-1]
        t     = inp[ip] if ip < len(inp) else '$'
        ti    = term_idx(t)
        stack_view = ' '.join(str(x) for x in st)

        if ti < 0:
            print(f"{step:<5} {stack_view:<27} {inp[ip:]:<10} error (invalid symbol)")
            return False

        act = action[state][ti]
        print(f"{step:<5} {stack_view:<27} {inp[ip:]:<10} ", end='')

        if act[0] == 's':
            nxt = int(act[1:])
            print(f"shift {nxt}")
            st.append(nxt)
            sym.append(t)
            ip += 1
        elif act[0] == 'r':
            p = int(act[1:])
            lhs, _ = prods[p]
            print(f"reduce {lhs} -> {''.join(prods[p][1])}")
            for _ in range(rhs_len[p]):
                st.pop(); sym.pop()
            g = gotos[st[-1]][nonterm_idx(lhs)]
            if g < 0:
                print("error (goto)")
                return False
            st.append(g)
            sym.append(lhs)
        elif act == 'acc':
            print("accept")
            return True
        else:
            print("error")
            return False
        step += 1

def main():
    print("Grammar:")
    print("E -> T T")
    print("T -> a T | b")
    print_table()
    raw = input("\nEnter input string over {a,b}: ").strip()
    print("\nResult:", "Accepted" if parse(raw) else "Rejected")

if __name__ == "__main__":
    main()
